{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ordered-carry",
   "metadata": {},
   "source": [
    "Scripts to create Supplementary_Table_3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "changing-leeds",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "# Load libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import json\n",
    "import signatureanalyzer as sa\n",
    "from glob import glob\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "genetic-skating",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define file paths\n",
    "DE_DIR = \"/home/yakiyama/DE_results/020222/\"\n",
    "\n",
    "HRD_HRP_DDR_DE_PATH = os.path.join(DE_DIR, \"HRD_v3/HRD_Global_02/analysis_files/DDR_diffexp_results_v3.tsv\")\n",
    "MMRD_FULL_DE_PATH = os.path.join(DE_DIR, \"MMRD/MMRDvsMMRP_excludePOLEPOLD/analysis_files/full_diffexp_results.tsv\")\n",
    "MMRD_DDR_DE_PATH = os.path.join(DE_DIR, \"MMRD/MMRDvsMMRP_excludePOLEPOLD/analysis_files/DDR_diffexp_results_v3.tsv\")\n",
    "MMRD_GSEA_PATH = os.path.join(DE_DIR, \"MMRD/MMRDvsMMRP_excludePOLEPOLD/analysis_files/full_gsea.tsv\")\n",
    "ACUTE_CHRONIC_FULL_DE_PATH = os.path.join(DE_DIR, \"HRD_v3/Group3Bvs2A/analysis_files/full_diffexp_results.tsv\")\n",
    "ACUTE_CHRONIC_DDR_DE_PATH = os.path.join(DE_DIR, \"HRD_v3/Group3Bvs2A/analysis_files/DDR_diffexp_results_v3.tsv\")\n",
    "ACUTE_CHRONIC_GSEA_PATH = os.path.join(DE_DIR, \"HRD_v3/Group3Bvs2A/analysis_files/full_gsea.tsv\")\n",
    "ACUTE_CHRONIC_PTMSEA_PATH = os.path.join(DE_DIR, \"HRD_v3/Group3Bvs2A/analysis_files/ptmsea_combined_results.tsv\")\n",
    "HRD_A_B_FULL_DE_PATH = os.path.join(\"DE_data/HRD_low_acute_hypoxia/analysis_files/full_diffexp_results.tsv\")\n",
    "HRD_A_B_GSEA_PATH = os.path.join(\"DE_data/HRD_low_acute_hypoxia/analysis_files/full_gsea.tsv\")\n",
    "HRD_A_C_FULL_DE_PATH = os.path.join(\"DE_data/HRD_low_chronic_hypoxia//analysis_files/full_diffexp_results.tsv\")\n",
    "HRD_A_C_GSEA_PATH = os.path.join(\"DE_data/HRD_low_chronic_hypoxia/analysis_files/full_gsea.tsv\")\n",
    "\n",
    "H_MUT_COLLAPSED_PATH = \"../results/mutational_signatures/Full_H_compressed_v3.tsv\"\n",
    "H_MUT_FULL_PATH = \"../results/mutational_signatures/H_uncollapsed_all.tsv\"\n",
    "W_MUT_FULL_PATH = \"../results/mutational_signatures/W_uncollapsed_all.tsv\"\n",
    "SAMPLE_MAP_PATH = \"gs://cptac-pancan-getzlab/mapping/PanCan_Participant_freeze_mapping.tsv\"\n",
    "SIGNATURE_SUMMARY_PATH = \"other_tables/Table_S3C_mutational_signatures_summary.tsv\"\n",
    "\n",
    "GENESET_PATH = \"/home/yakiyama/DE_data/genesets/092121/full_geneset_v3.json\"\n",
    "\n",
    "HRD_HRP_CP_DATA_PATH = \"/home/yakiyama/CPTAC_PanCan_2021/supplemental_data/causalpath/HRD_Global_02_data.txt\"\n",
    "HRD_HRP_CP_PARAM_PATH = \"/home/yakiyama/CPTAC_PanCan_2021/supplemental_data/causalpath/HRD_Global_02_parameters.txt\"\n",
    "ACUTE_CP_DATA_PATH = \"/home/yakiyama/CPTAC_PanCan_2021/supplemental_data/causalpath/Group3Bvs2A_data.txt\"\n",
    "ACUTE_CP_PARAM_PATH = \"/home/yakiyama/CPTAC_PanCan_2021/supplemental_data/causalpath/Group3Bvs2A_parameters.txt\"\n",
    "MMRD_MMRP_CP_DATA_PATH = \"/home/yakiyama/CPTAC_PanCan_2021/supplemental_data/causalpath/MMRD_data.txt\"\n",
    "MMRD_MMRP_CP_PARAM_PATH = \"/home/yakiyama/CPTAC_PanCan_2021/supplemental_data/causalpath/MMRD_parameters.txt\"\n",
    "\n",
    "KINASE_LIBRARY_PATH = \"/home/yakiyama/CPTAC_PanCan_2021/analysis_files/kinase_library_051022/2.Dendrogram/phosphoproteome_res/enrichment_results\"\n",
    "\n",
    "MMRD_FISHER_PATH = \"other_tables/Table_S3N_MMRD_fisher.tsv\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "powered-treasure",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Symbols</th>\n",
       "      <th>Sites</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Effect</th>\n",
       "      <th>HRD_data</th>\n",
       "      <th>HRD_hypoxia_data</th>\n",
       "      <th>MMRD_data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KDM4A-P-S410</th>\n",
       "      <td>KDM4A</td>\n",
       "      <td>S410</td>\n",
       "      <td>P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.770617</td>\n",
       "      <td>0.440954</td>\n",
       "      <td>-0.602466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PARP4-P-S1434</th>\n",
       "      <td>PARP4</td>\n",
       "      <td>S1434</td>\n",
       "      <td>P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.766541</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLD3-P-T277</th>\n",
       "      <td>POLD3</td>\n",
       "      <td>T277</td>\n",
       "      <td>P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010874</td>\n",
       "      <td>0.084983</td>\n",
       "      <td>0.722302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRWD1-R</th>\n",
       "      <td>LRWD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.996278</td>\n",
       "      <td>0.646388</td>\n",
       "      <td>-0.793394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BAZ1B-P-S330</th>\n",
       "      <td>BAZ1B</td>\n",
       "      <td>S330</td>\n",
       "      <td>P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.015007</td>\n",
       "      <td>0.059176</td>\n",
       "      <td>0.988916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAXIP1-P-S227</th>\n",
       "      <td>PAXIP1</td>\n",
       "      <td>S227</td>\n",
       "      <td>P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.602466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABRAXAS1-P-S387-T390</th>\n",
       "      <td>ABRAXAS1</td>\n",
       "      <td>S387|T390</td>\n",
       "      <td>P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.894740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLE-P-S1297</th>\n",
       "      <td>POLE</td>\n",
       "      <td>S1297</td>\n",
       "      <td>P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.106377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL1-P-S917</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>S917</td>\n",
       "      <td>P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.935834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC6-P-S486</th>\n",
       "      <td>ERCC6</td>\n",
       "      <td>S486</td>\n",
       "      <td>P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.840129</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3776 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Symbols      Sites Feature  Effect  HRD_data  \\\n",
       "KDM4A-P-S410             KDM4A       S410       P     NaN -0.770617   \n",
       "PARP4-P-S1434            PARP4      S1434       P     NaN  0.766541   \n",
       "POLD3-P-T277             POLD3       T277       P     NaN  0.010874   \n",
       "LRWD1-R                  LRWD1        NaN       R     NaN -0.996278   \n",
       "BAZ1B-P-S330             BAZ1B       S330       P     NaN  0.015007   \n",
       "...                        ...        ...     ...     ...       ...   \n",
       "PAXIP1-P-S227           PAXIP1       S227       P     NaN       NaN   \n",
       "ABRAXAS1-P-S387-T390  ABRAXAS1  S387|T390       P     NaN       NaN   \n",
       "POLE-P-S1297              POLE      S1297       P     NaN       NaN   \n",
       "ABL1-P-S917               ABL1       S917       P     NaN       NaN   \n",
       "ERCC6-P-S486             ERCC6       S486       P     NaN       NaN   \n",
       "\n",
       "                      HRD_hypoxia_data  MMRD_data  \n",
       "KDM4A-P-S410                  0.440954  -0.602466  \n",
       "PARP4-P-S1434                      NaN        NaN  \n",
       "POLD3-P-T277                  0.084983   0.722302  \n",
       "LRWD1-R                       0.646388  -0.793394  \n",
       "BAZ1B-P-S330                  0.059176   0.988916  \n",
       "...                                ...        ...  \n",
       "PAXIP1-P-S227                      NaN   0.602466  \n",
       "ABRAXAS1-P-S387-T390               NaN   0.894740  \n",
       "POLE-P-S1297                       NaN   0.106377  \n",
       "ABL1-P-S917                        NaN   0.935834  \n",
       "ERCC6-P-S486                       NaN  -0.840129  \n",
       "\n",
       "[3776 rows x 7 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combined CausalPath Data\n",
    "hrd_hrp_cp_df = pd.read_csv(HRD_HRP_CP_DATA_PATH, sep='\\t', index_col=0)\n",
    "hypoxia_cp_df = pd.read_csv(ACUTE_CP_DATA_PATH, sep='\\t', index_col=0)\n",
    "mmrd_mmrp_cp_df = pd.read_csv(MMRD_MMRP_CP_DATA_PATH, sep='\\t', index_col=0)\n",
    "\n",
    "merged_cp_df = pd.concat([hrd_hrp_cp_df, hypoxia_cp_df, mmrd_mmrp_cp_df], 1)\n",
    "merged_cp_df['Sites_Collapsed'] = merged_cp_df.apply(lambda x: x['Sites'].dropna().unique()[0] if '-P-' in x.name else np.nan, 1)\n",
    "merged_cp_df['Effect_Collapsed'] = merged_cp_df.apply(lambda x: x['Effect'].unique()[0], 1)\n",
    "merged_cp_df['Symbols_Collapsed'] = merged_cp_df.apply(lambda x: x['Symbols'].dropna().unique()[0], 1)\n",
    "merged_cp_df['Feature_Collapsed'] = merged_cp_df.apply(lambda x: x['Feature'].dropna().unique()[0], 1)\n",
    "\n",
    "merged_cp_df = merged_cp_df.drop(columns=['Sites'])\n",
    "merged_cp_df = merged_cp_df.drop(columns=['Effect'])\n",
    "merged_cp_df = merged_cp_df.drop(columns=['Symbols'])\n",
    "merged_cp_df = merged_cp_df.drop(columns=['Feature'])\n",
    "\n",
    "merged_cp_df = merged_cp_df.rename(columns={\"Sites_Collapsed\":\"Sites\", \n",
    "                                            \"Effect_Collapsed\":\"Effect\", \n",
    "                                            \"Feature_Collapsed\":\"Feature\", \n",
    "                                            \"Symbols_Collapsed\":\"Symbols\"})\n",
    "\n",
    "merged_cp_df = merged_cp_df[['Symbols', 'Sites', 'Feature', 'Effect', 'HRD', 'HRD3Bvs2A_3B', 'MMRDvsMMRP_MMRD']]\n",
    "merged_cp_df = merged_cp_df.rename(columns={'HRD3Bvs2A_3B':'HRD_hypoxia_data', \n",
    "                                            'HRD':'HRD_data', \n",
    "                                            'MMRDvsMMRP_MMRD':'MMRD_data'})\n",
    "merged_cp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "irish-policy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HRD_params</th>\n",
       "      <th>HRD_hypoxia_params</th>\n",
       "      <th>MMRD_params</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>proteomics-values-file = ../../data.txt</td>\n",
       "      <td>proteomics-values-file = ../../data.txt</td>\n",
       "      <td>proteomics-values-file = ../../data.txt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>id-column = ID</td>\n",
       "      <td>id-column = ID</td>\n",
       "      <td>id-column = ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>symbols-column = Symbols</td>\n",
       "      <td>symbols-column = Symbols</td>\n",
       "      <td>symbols-column = Symbols</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sites-column = Sites</td>\n",
       "      <td>sites-column = Sites</td>\n",
       "      <td>sites-column = Sites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>feature-column = Feature</td>\n",
       "      <td>feature-column = Feature</td>\n",
       "      <td>feature-column = Feature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>effect-column = Effect</td>\n",
       "      <td>effect-column = Effect</td>\n",
       "      <td>effect-column = Effect</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>value-transformation = signed-p-values</td>\n",
       "      <td>value-transformation = signed-p-values</td>\n",
       "      <td>value-transformation = signed-p-values</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>threshold-for-data-significance = 0.105 protein</td>\n",
       "      <td>threshold-for-data-significance = 0.105 protein</td>\n",
       "      <td>threshold-for-data-significance = 0.105 protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>threshold-for-data-significance = 0.105 phosph...</td>\n",
       "      <td>threshold-for-data-significance = 0.105 phosph...</td>\n",
       "      <td>threshold-for-data-significance = 0.105 phosph...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>threshold-for-data-significance = 0.105 acetyl...</td>\n",
       "      <td>threshold-for-data-significance = 0.105 acetyl...</td>\n",
       "      <td>threshold-for-data-significance = 0.105 acetyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>threshold-for-data-significance = 0.105 rna</td>\n",
       "      <td>threshold-for-data-significance = 0.105 rna</td>\n",
       "      <td>threshold-for-data-significance = 0.105 rna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>color-saturation-value = 15</td>\n",
       "      <td>color-saturation-value = 15</td>\n",
       "      <td>color-saturation-value = 15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>calculate-network-significance = true</td>\n",
       "      <td>calculate-network-significance = true</td>\n",
       "      <td>calculate-network-significance = true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>permutations-for-significance = 10000</td>\n",
       "      <td>permutations-for-significance = 10000</td>\n",
       "      <td>permutations-for-significance = 10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>fdr-threshold-for-network-significance = 0.1</td>\n",
       "      <td>fdr-threshold-for-network-significance = 0.1</td>\n",
       "      <td>fdr-threshold-for-network-significance = 0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>use-network-significance-for-causal-reasoning ...</td>\n",
       "      <td>use-network-significance-for-causal-reasoning ...</td>\n",
       "      <td>use-network-significance-for-causal-reasoning ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>show-all-genes-with-proteomic-data = true</td>\n",
       "      <td>show-all-genes-with-proteomic-data = true</td>\n",
       "      <td>show-all-genes-with-proteomic-data = true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>relation-filter-type = site-specific-only</td>\n",
       "      <td>relation-filter-type = site-specific-only</td>\n",
       "      <td>relation-filter-type = site-specific-only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>value-column = HRD</td>\n",
       "      <td>value-column = acute_hypoxia</td>\n",
       "      <td>value-column = MMRDvsMMRP_MMRD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           HRD_params  \\\n",
       "0             proteomics-values-file = ../../data.txt   \n",
       "1                                      id-column = ID   \n",
       "2                            symbols-column = Symbols   \n",
       "3                                sites-column = Sites   \n",
       "4                            feature-column = Feature   \n",
       "5                              effect-column = Effect   \n",
       "6              value-transformation = signed-p-values   \n",
       "7     threshold-for-data-significance = 0.105 protein   \n",
       "8   threshold-for-data-significance = 0.105 phosph...   \n",
       "9   threshold-for-data-significance = 0.105 acetyl...   \n",
       "10        threshold-for-data-significance = 0.105 rna   \n",
       "11                        color-saturation-value = 15   \n",
       "12              calculate-network-significance = true   \n",
       "13              permutations-for-significance = 10000   \n",
       "14       fdr-threshold-for-network-significance = 0.1   \n",
       "15  use-network-significance-for-causal-reasoning ...   \n",
       "16          show-all-genes-with-proteomic-data = true   \n",
       "17          relation-filter-type = site-specific-only   \n",
       "18                                 value-column = HRD   \n",
       "\n",
       "                                   HRD_hypoxia_params  \\\n",
       "0             proteomics-values-file = ../../data.txt   \n",
       "1                                      id-column = ID   \n",
       "2                            symbols-column = Symbols   \n",
       "3                                sites-column = Sites   \n",
       "4                            feature-column = Feature   \n",
       "5                              effect-column = Effect   \n",
       "6              value-transformation = signed-p-values   \n",
       "7     threshold-for-data-significance = 0.105 protein   \n",
       "8   threshold-for-data-significance = 0.105 phosph...   \n",
       "9   threshold-for-data-significance = 0.105 acetyl...   \n",
       "10        threshold-for-data-significance = 0.105 rna   \n",
       "11                        color-saturation-value = 15   \n",
       "12              calculate-network-significance = true   \n",
       "13              permutations-for-significance = 10000   \n",
       "14       fdr-threshold-for-network-significance = 0.1   \n",
       "15  use-network-significance-for-causal-reasoning ...   \n",
       "16          show-all-genes-with-proteomic-data = true   \n",
       "17          relation-filter-type = site-specific-only   \n",
       "18                       value-column = acute_hypoxia   \n",
       "\n",
       "                                          MMRD_params  \n",
       "0             proteomics-values-file = ../../data.txt  \n",
       "1                                      id-column = ID  \n",
       "2                            symbols-column = Symbols  \n",
       "3                                sites-column = Sites  \n",
       "4                            feature-column = Feature  \n",
       "5                              effect-column = Effect  \n",
       "6              value-transformation = signed-p-values  \n",
       "7     threshold-for-data-significance = 0.105 protein  \n",
       "8   threshold-for-data-significance = 0.105 phosph...  \n",
       "9   threshold-for-data-significance = 0.105 acetyl...  \n",
       "10        threshold-for-data-significance = 0.105 rna  \n",
       "11                        color-saturation-value = 15  \n",
       "12              calculate-network-significance = true  \n",
       "13              permutations-for-significance = 10000  \n",
       "14       fdr-threshold-for-network-significance = 0.1  \n",
       "15  use-network-significance-for-causal-reasoning ...  \n",
       "16          show-all-genes-with-proteomic-data = true  \n",
       "17          relation-filter-type = site-specific-only  \n",
       "18                     value-column = MMRDvsMMRP_MMRD  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hrd_hrp_param_df = pd.read_csv(HRD_HRP_CP_PARAM_PATH, sep='\\t', header=None)\n",
    "hypoxia_param_df = pd.read_csv(ACUTE_CP_PARAM_PATH, sep='\\t', header=None)\n",
    "mmrd_param_df = pd.read_csv(MMRD_MMRP_CP_PARAM_PATH, sep='\\t', header=None)\n",
    "\n",
    "param_df = pd.concat([hrd_hrp_param_df ,hypoxia_param_df, mmrd_param_df], 1)\n",
    "param_df.columns = ['HRD_params', 'HRD_hypoxia_params', 'MMRD_params']\n",
    "param_df.loc[18, \"HRD_hypoxia_params\"] = 'value-column = acute_hypoxia'\n",
    "param_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "specific-aquatic",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yakiyama/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3072: DtypeWarning: Columns (13,14,15) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# Mutational signature data\n",
    "H_mut_full_df = pd.read_csv(H_MUT_FULL_PATH, sep='\\t', index_col=0)\n",
    "W_mut_df = pd.read_csv(W_MUT_FULL_PATH, sep='\\t', index_col=0)\n",
    "H_mut_collapsed_df = pd.read_csv(H_MUT_COLLAPSED_PATH, sep='\\t', index_col=0)\n",
    "## Remove columns from HRD signatures that included UV sample\n",
    "H_mut_full_df = H_mut_full_df[[x for x in H_mut_full_df.columns if 'UV' not in x]]\n",
    "W_mut_df = W_mut_df[[x for x in W_mut_df.columns if 'UV' not in x]]\n",
    "\n",
    "# Global DDR analyses\n",
    "hrd_hrp_ddr_df = pd.read_csv(HRD_HRP_DDR_DE_PATH, sep='\\t', index_col=0)\n",
    "mmrd_ddr_df = pd.read_csv(MMRD_DDR_DE_PATH, sep='\\t', index_col=0)\n",
    "mmrd_gsea_df = pd.read_csv(MMRD_GSEA_PATH, sep='\\t', index_col=0)\n",
    "\n",
    "# HRD Subtype analyses\n",
    "acute_full_de_df = pd.read_csv(ACUTE_CHRONIC_FULL_DE_PATH, sep='\\t', index_col=0)\n",
    "acute_ddr_de_df = pd.read_csv(ACUTE_CHRONIC_DDR_DE_PATH, sep='\\t', index_col=0)\n",
    "acute_chronic_gsea_df = pd.read_csv(ACUTE_CHRONIC_GSEA_PATH, sep='\\t', index_col=0)\n",
    "acute_ptmsea_df = pd.read_csv(ACUTE_CHRONIC_PTMSEA_PATH, sep='\\t', index_col=0)\n",
    "hrd_a_b_full_de_df = pd.read_csv(HRD_A_B_FULL_DE_PATH, sep='\\t', index_col=0)\n",
    "hrd_a_b_gsea_df = pd.read_csv(HRD_A_B_GSEA_PATH, sep='\\t', index_col=0)\n",
    "hrd_a_c_full_de_df = pd.read_csv(HRD_A_C_FULL_DE_PATH, sep='\\t', index_col=0)\n",
    "hrd_a_c_gsea_df = pd.read_csv(HRD_A_C_GSEA_PATH, sep='\\t', index_col=0)\n",
    "\n",
    "# Rename acute-chronic IDs\n",
    "acute_full_de_df['id'] = acute_full_de_df['id'].map({'HRD3Bvs2A_2A':'chronic_hypoxia', \n",
    "                                                     'HRD3Bvs2A_3B':'acute_hypoxia'})\n",
    "acute_ddr_de_df['id'] = acute_ddr_de_df['id'].map({'HRD3Bvs2A_3B':'acute_hypoxia'})\n",
    "acute_chronic_gsea_df['id'] = acute_chronic_gsea_df['id'].map({'HRD3Bvs2A_3B':'acute_hypoxia'})\n",
    "acute_ptmsea_df['id'] = acute_ptmsea_df['id'].map({'HRD3Bvs2A_3B':'acute_hypoxia'})\n",
    "\n",
    "# MMRD analyses\n",
    "mmrd_full_de_df = pd.read_csv(MMRD_FULL_DE_PATH, sep='\\t', index_col=0)\n",
    "mmrd_ddr_de_df = pd.read_csv(MMRD_DDR_DE_PATH, sep='\\t', index_col=0)\n",
    "mmrd_gsea_df = pd.read_csv(MMRD_GSEA_PATH, sep='\\t', index_col=0)\n",
    "mmrd_full_de_df['id'] = mmrd_full_de_df['id'].map({'MMRDvsMMRP_MMRD':'MMRD', 'MMRDvsMMRP_MMRP':'MMRP'})\n",
    "mmrd_ddr_de_df['id'] = mmrd_ddr_de_df['id'].map({'MMRDvsMMRP_MMRD':'MMRD', 'MMRDvsMMRP_MMRP':'MMRP'})\n",
    "mmrd_gsea_df['id'] = mmrd_gsea_df['id'].map({'MMRDvsMMRP_MMRD':'MMRD', 'MMRDvsMMRP_MMRP':'MMRP'})\n",
    "\n",
    "# Geneset\n",
    "gs_d = {}\n",
    "with open(GENESET_PATH, \"r\") as oFile:\n",
    "    gs_d = json.load(oFile)\n",
    "ddr_gs_df = pd.DataFrame(pd.Series(gs_d['DDR']), columns=['Genes'])\n",
    "\n",
    "# PTM-SEA Subset for just one side of the comparison\n",
    "acute_ptmsea_df = acute_ptmsea_df[acute_ptmsea_df['id'] == 'acute_hypoxia']\n",
    "\n",
    "# Mutational signatures: standardize sample IDs to Proteome_Sample_ID\n",
    "sample_df = pd.read_csv(SAMPLE_MAP_PATH, sep='\\t')\n",
    "H_mut_full_df.index = H_mut_full_df.index.map(sample_df.set_index('WXS').to_dict()['Proteome_Sample_ID'])\n",
    "H_mut_collapsed_df.index = H_mut_collapsed_df.index.map(sample_df.set_index('WXS').to_dict()['Proteome_Sample_ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "numerous-airport",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Repair Deficiency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C01CO014</th>\n",
       "      <td>MMRD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C01CO022</th>\n",
       "      <td>MMRD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X05BR029</th>\n",
       "      <td>MMRD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C05CO003</th>\n",
       "      <td>MMRD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C05CO006</th>\n",
       "      <td>MMRD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C26OV011</th>\n",
       "      <td>HRD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L.02613.T</th>\n",
       "      <td>HRD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N.03069.T</th>\n",
       "      <td>HRD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N.03426.T</th>\n",
       "      <td>HRD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N.03430.T</th>\n",
       "      <td>HRD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>145 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Repair Deficiency\n",
       "C01CO014                 MMRD\n",
       "C01CO022                 MMRD\n",
       "X05BR029                 MMRD\n",
       "C05CO003                 MMRD\n",
       "C05CO006                 MMRD\n",
       "...                       ...\n",
       "C26OV011                  HRD\n",
       "C3L.02613.T               HRD\n",
       "C3N.03069.T               HRD\n",
       "C3N.03426.T               HRD\n",
       "C3N.03430.T               HRD\n",
       "\n",
       "[145 rows x 1 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DNA repair deficiency map\n",
    "ddr_classify_df = H_mut_collapsed_df.copy()\n",
    "ddr_classify_df = ddr_classify_df[(ddr_classify_df['HRD']>45) | (ddr_classify_df['MMRD']>0)]\n",
    "ddr_classify_df['Repair Deficiency'] = ddr_classify_df.apply(lambda x: 'MMRD' if x['MMRD']>0 else 'HRD', 1)\n",
    "ddr_classify_df = ddr_classify_df[['Repair Deficiency']]\n",
    "ddr_classify_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "according-struggle",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mutational signatures summary table\n",
    "signature_summary_df = pd.read_csv(SIGNATURE_SUMMARY_PATH, sep='\\t', index_col=0)\n",
    "\n",
    "# MMRD MS-indel summary statistics\n",
    "mmrd_fisher_df = pd.read_csv(MMRD_FISHER_PATH, sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "present-singing",
   "metadata": {},
   "outputs": [],
   "source": [
    "options = {}\n",
    "options['strings_to_formulas'] = False\n",
    "options['strings_to_urls'] = False\n",
    "\n",
    "with pd.ExcelWriter(\"Supplemental_Table_3_v5.xlsx\", endgine='xlsxwriter', options=options) as excel_writer:\n",
    "    W_mut_df.to_excel(excel_writer, sheet_name='Table 3A')\n",
    "    H_mut_full_df.to_excel(excel_writer, sheet_name='Table 3B')\n",
    "    H_mut_collapsed_df.to_excel(excel_writer, sheet_name='Table 3C')\n",
    "    signature_summary_df.to_excel(excel_writer, sheet_name='Table 3D')\n",
    "    ddr_classify_df.to_excel(excel_writer, sheet_name='Table 3E')\n",
    "    \n",
    "    # Geneset\n",
    "    ddr_gs_df.to_excel(excel_writer, sheet_name='Table 3F')\n",
    "\n",
    "    # Global HRD DDR Analysis\n",
    "    hrd_hrp_ddr_df.to_excel(excel_writer, sheet_name='Table 3G')\n",
    "\n",
    "    # HRD Subtype analyses\n",
    "    acute_full_de_df.to_excel(excel_writer, sheet_name='Table 3H')\n",
    "    acute_ddr_de_df.to_excel(excel_writer, sheet_name='Table 3I')\n",
    "    acute_chronic_gsea_df.to_excel(excel_writer, sheet_name='Table 3J')\n",
    "    acute_ptmsea_df.to_excel(excel_writer, sheet_name='Table 3K')\n",
    "    hrd_a_b_full_de_df.to_excel(excel_writer, sheet_name='Table 3L')\n",
    "    hrd_a_b_gsea_df.to_excel(excel_writer, sheet_name='Table 3M')\n",
    "    hrd_a_c_full_de_df.to_excel(excel_writer, sheet_name='Table 3N')\n",
    "    hrd_a_c_gsea_df.to_excel(excel_writer, sheet_name='Table 3O')\n",
    "    # Table S3P Kinase Library causal path scores\n",
    "    \n",
    "    \n",
    "    # MMRD analyses\n",
    "    mmrd_full_de_df.to_excel(excel_writer, sheet_name='Table 3Q')\n",
    "    mmrd_ddr_df.to_excel(excel_writer, sheet_name='Table 3R')\n",
    "    mmrd_gsea_df.to_excel(excel_writer, sheet_name='Table 3S')\n",
    "\n",
    "    # MMRD Fisher tests\n",
    "    mmrd_fisher_df.to_excel(excel_writer, sheet_name='Table 3T')\n",
    "\n",
    "    excel_writer.save()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bigger-violin",
   "metadata": {},
   "source": [
    "## Kinase Library results for Supplemental Table 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "adjusted-mitchell",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_dendro_df = pd.DataFrame()\n",
    "for dendro_file in glob(os.path.join(KINASE_LIBRARY_PATH, \"*tsv\")):\n",
    "    dendrosplit_id = int(dendro_file.split('/')[-1].split('.')[0])\n",
    "    dendro_df = pd.read_csv(dendro_file, sep='\\t')\n",
    "    dendro_df['dendrogram_split'] = dendrosplit_id\n",
    "    full_dendro_df = pd.concat([full_dendro_df, dendro_df])\n",
    "full_dendro_df.rename(columns={\"Unnamed: 0\":\"geneSymbol\"}, inplace=True)\n",
    "full_dendro_df.sort_values(\"dendrogram_split\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "subtle-ridge",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_dendro_df.to_csv(\"../analysis_files/kinase_library_051022/full_dendrogram_KinaseLibrary_results.tsv\", \n",
    "                      sep='\\t')\n",
    "options = {}\n",
    "options['strings_to_formulas'] = False\n",
    "options['strings_to_urls'] = False\n",
    "with pd.ExcelWriter(\"../analysis_files/kinase_library_051022/full_dendrogram_KinaseLibrary_results.xlsx\", \n",
    "                    endgine='xlsxwriter', options=options) as excel_writer:\n",
    "    full_dendro_df.to_excel(excel_writer, sheet_name=\"KinaseLibrary\")\n",
    "    excel_writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "native-yemen",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
